platform/security/idp/security-idp-plugin/security-idp-plugin-nameidpolicy/src/main/java/org/codice/ddf/security/idp/plugin/nameidpolicy/NameIdPolicyPresignPlugin.java

```diff
@@ -0,0 +1,136 @@
+ * according to the values found in the NameIDPolicy of the passed in AuthnRequest.
+ */
+public class NameIdPolicyPresignPlugin implements SamlPresignPlugin {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(NameIdPolicyPresignPlugin.class);
+
+  private static final String EMAIL_ATTRIBUTE_NAME =
+      "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress";
+
+  @Override
+  public void processPresign(
+      Response response,
+      AuthnRequest authnRequest,
+      List<String> spMetadata,
+      Set<Binding> supportedBindings) {
+    NameIDPolicy nameIdPolicy = authnRequest.getNameIDPolicy();
+
+    if (nameIdPolicy == null
+        || (StringUtils.isEmpty(nameIdPolicy.getFormat())
+        && StringUtils.isEmpty(nameIdPolicy.getSPNameQualifier()))) {
+      return;
+    }
+
+    response
+        .getAssertions()
+        .stream()
+        .map(assertion -> assertion.getSubject().getNameID())
+        .forEach(
+            assertionNameId ->
+                setAssertionNameIdQualifierAndFormat(
+                    assertionNameId,
+                    nameIdPolicy.getFormat(),
+                    nameIdPolicy.getSPNameQualifier(),
+                    response));
+  }
+
+  private void setAssertionNameIdQualifierAndFormat(
+      NameID assertionNameId,
+      @Nullable String nameIdFormatPolicy,
+      @Nullable String spNameQualifierPolicy,
+      Response response) {
+
+    if (StringUtils.isNotEmpty(spNameQualifierPolicy)) {
+      assertionNameId.setSPNameQualifier(spNameQualifierPolicy);
+    }
+
+    if (StringUtils.isNotEmpty(nameIdFormatPolicy)) {
+      switch (nameIdFormatPolicy) {
+        // supported NameIDFormats
+        case NameID.UNSPECIFIED:
+          break;
+        case NameID.PERSISTENT:
+          // TODO DDF-3965
+          break;
+        case NameID.TRANSIENT:
+          // TODO DDF-3965
+          break;
+
+        // partially supported NameIDFormats
+        case NameID.X509_SUBJECT:
+          LOGGER.warn("Specifying the \"X509Subject\" NameIDPolicy Format is not fully supported.");
+          if (!NameID.X509_SUBJECT.equals(assertionNameId.getFormat())) {
+            throw new UnsupportedOperationException(
+                "The \"X509Subject\" NameID could not be retrieved.");
+          }
+          break;
+        case NameID.EMAIL:
+          LOGGER.warn("Specifying the \"Email\" NameIDPolicy Format is not fully supported.");
+          assertionNameId.setValue(resolveEmail(response));
+          break;
+
+        // not supported NameIDFormats
+        case NameID.WIN_DOMAIN_QUALIFIED:
+        case NameID.KERBEROS:
+        case NameID.ENTITY:
+        default:
+          throw new UnsupportedOperationException(
+              String.format(
+                  "The NameIDPolicy Format of value %s is not supported.", nameIdFormatPolicy));
+      }
+    }
+
+    assertionNameId.setFormat(nameIdFormatPolicy);
```

Comment:
❗️ Don't we want to fail if the format value already exists and it's not equal to the policy's format? For example, I think it's incorrect for us to have an email formatted nameID, but to then overwrite the format value to persistent.

Reply:
In our case, and probably many others, the Subject's NameID is only there to say "by the way, this is how I am going to refer to the user/subject/principal". Depending on how the relying SP decides to use it, it can lead to errors. Luckily for us, our SP only uses the Subject's NameID to store it as the user's "principal".

That being said, our SP parses that NameID based on the Format value, and can error out if we get a NameID whose value and format don't match up. I.e. trying to parse an email as a certificate can get messy. In this PR that is mitigated because we actually do what you said for the `X509SubjectName` NameIDPolicy.

TLDR: It isn't broken, but I can change it if that is our desired behavior.

Reply:
I think we're only doing that for if `X509SubjectName` is the format on the NameIDPolicy, but what if `X509SubjectName` is the format of the nameID itself and `persistent` is the format of the policy. We will then change the format to `persistent` even though really the NameID has a `X509SubjectName` value.

Reply:
Yes, we would be sending an `X509SubjectName` as a  `persistent` identifier. However, DDF's SP doesn't care what String value it gets for the `persistent` identifier. It's only really picky about what String value is in the `X509SubjectName` identifier.